### PR TITLE
Extending monitoring upgrade test

### PR DESCRIPTION
**Pull request by shorim**
**Description**

Changes proposed in this pull request:

- Check that all targets are healthy before and after upgrade. Checking for target with job=istio-ingressgateway is excluded for now, because it is not healthy in release 1.11.1 and this check will be added in the next release
- Check that all rules are healthy before and after upgrade

**Related issue(s)**

#4501

### CODE REVIEW

File: tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+		return err
+	}
+	return nil
+}
+
+// CreateResources checks that all targets and rules are healthy after upgrade
+func (t TargetsAndRulesTest) TestResources(stop <-chan struct{}, log logrus.FieldLogger, namespace string) error {
+	log.Println("checking that monitoring pods are ready")
+	err := t.testPodsAreReady()
+	if err != nil {
```
Comment (clebs) [inline tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go:65]:
```suggestion
	if err := t.testPodsAreReady(); err != nil {
```

File: tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+		return err
+	}
+	log.Println("checking that all targets are healthy after upgrade")
+	err = t.testTargetsAreHealthy()
+	if err != nil {
```
Comment (clebs) [inline tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go:70]:
```suggestion
	if err := t.testTargetsAreHealthy(); err != nil {
```

File: tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+		return err
+	}
+	log.Println("checking that all rules are healthy after upgrade")
+	err = t.testRulesAreHealthy()
+	if err != nil {
```
Comment (clebs) [inline tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go:75]:
```suggestion
	if err := t.testRulesAreHealthy(); err != nil {
```

File: tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+			return errors.Errorf(timeoutMessage)
+		case <-tick.C:
+			var resp promAPI.TargetsResponse
+			url := fmt.Sprintf("%s/api/v1/targets", prometheusURL)
+			respBody, statusCode, err := t.doGet(url)
+			if err != nil {
+				return errors.Wrap(err, "cannot query targets")
+			}
+			err = json.Unmarshal([]byte(respBody), &resp)
+			if err != nil {
```
Comment (clebs) [inline tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go:176]:
```suggestion
			err := json.Unmarshal([]byte(respBody), &resp); err != nil {
```

File: tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+			return errors.Errorf(timeoutMessage)
+		case <-tick.C:
+			var resp promAPI.AlertResponse
+			url := fmt.Sprintf("%s/api/v1/rules", prometheusURL)
+			respBody, statusCode, err := t.doGet(url)
+			if err != nil {
+				return errors.Wrap(err, "cannot query rules")
+			}
+			err = json.Unmarshal([]byte(respBody), &resp)
+			if err != nil {
```
Comment (clebs) [inline tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go:216]:
```suggestion
			if err := json.Unmarshal([]byte(respBody), &resp); err != nil {
```

File: tests/end-to-end/upgrade/pkg/tests/monitoring/promAPI/targets.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package promAPI
```
[the remark this answers is not in the document]
Comment (shorim) [reply tests/end-to-end/upgrade/pkg/tests/monitoring/promAPI/targets.go]: package is renamed now to `prom`

File: tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+			if expectedAlertManagers == actualAlertManagers && expectedNodeExporter == actualNodeExporter && expectedPrometheusInstances == actualPrometheusInstances && expectedKubeStateMetrics == actualKubeStateMetrics && expectedGrafanaInstance == actualGrafanaInstance {
+				return nil
+			}
+		}
+	}
+}
+
+func (t TargetsAndRulesTest) testTargetsAreHealthy() error {
+	timeout := time.After(3 * time.Minute)
+	tick := time.NewTicker(30 * time.Second)
```
Comment (piotrmiskiewicz) [inline tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go:155]: do we need to wait 30 seconds for the first check?

**Comment by piotrmiskiewicz** [comment]
General question: how log does it take the "testResource" method usually?

File: tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go (hunk 7 above)
Comment (shorim) [reply tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go:155]: @piotrmiskiewicz the idea is that the scraping interval is 30 seconds, so if in the first tick one of the targets was down, then there is no meaning to check again in less than 30 seconds beacause promtheus would have not yet done another scraping and the check will still fail

Comment (shorim) [reply tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go:155]: but at the same time you are right, this makes the first check only happens after 30 seconds, so in total that makes the test runs in around 1 minute

Comment (piotrmiskiewicz) [reply tests/end-to-end/upgrade/pkg/tests/monitoring/targets_rules.go:155]: so please do the first check without waiting 30 seconds

**Pull request merged**